### DISCUSSION TITLE
Copy file SD to USB (fatfs)

### DISCUSSION THREAD

**Discussion by arsamus** [rsta2/circle#457] [Development]
Hi. I'm implementing a file copy method with fatfs library for a Sampler project based on MiniSynth class structure. It seams to work ok when I copy files from USB to SD, but, on the other wise, when I've tried to copy files from SD to USB the system (or maybe the code) hungs after create the file (f_open) (zero size file is created) SD usb USB drives are correctly mounted on kernel's run method and the file copy method is kicked from kernel's main loop.

What thing I'm missing or doing wrong ? Here is the code. Is amost the same as the posted on fatfs project page:
```
bool CSynthConfig::CopyFile(const char *source, const char *target)

{

	FIL File1, File2;
FRESULT	Result;
	BYTE buffer[4096];  
	unsigned br, bw;

	Result = f_open(&File1, source, FA_READ);

	if (Result != FR_OK)
	{
		return 0;
	}

	Result = f_open(&File2, target, FA_WRITE | FA_CREATE_ALWAYS);

	if (Result != FR_OK)
	{
		f_close(&File1);
		return 0;
	}	

	bool e = 1;
    for (;;) 
	{
        Result = f_read(&File1, buffer, sizeof buffer, &br);  /* Read a chunk of source file */
		if(Result)
		{
			e = 0;
			break;
		}

		if (br == 0) break; // eof
        Result = f_write(&File2, buffer, br, &bw);            /* Write it to the destination file */
        if (Result || bw < br) /* error or disk full */
		{

			e = 0;
			break;

		}

    }

	e &= (f_size(&File1) == f_size(&File2)); 

    f_close(&File1);
    f_close(&File2);
	return e;

}
```

**Comment by rsta2** [comment]
I tried your `CopyFile()` on a Raspberry Pi 5 with the attached test program, but unfortunately I cannot reproduce the problem. The program does work. If the problem remains with this program at your site, I need to know, which Raspberry Pi model you are using and in which mode (32 or 64-bits).

[copyfile.zip](https://github.com/user-attachments/files/15899512/copyfile.zip)

**Reply by arsamus** [reply]
After several days of wracking my brain without finding a propper solution, rewritting different alternatives, I just realized that the problem was the USB pen drive... Since it was new, I dismissed the possibility that it had issues, but indeed, it did. I tried others and they worked fine. Thank you very much anyway for testing the code; it helped me rule out that the problem was there and suspect about the hardware!

**Reply by rsta2** [reply]
You are welcome. No, problem! It's good that you have found the reason.
